src/integration.ts

```diff
@@ -301,20 +302,143 @@ export function makeIntegration<K extends string, F extends AnyFunction>(
+                new ReferenceExpr("", () => integration),
+                node.args.map((arg) => arg.clone())
+              )
+            )
+          )
+        );
+      }
     }
-    return visitEachChild(node, find);
+
+    return visitEachChild(node, visit);
   });
-  return integrations;
+
+  return nodes;
+}
+
+/**
+ * A bottom-up algorithm that determines the ONLY {@link Integration} value that the {@link node}
+ * will resolve to at runtime.
+ *
+ * If the {@link node} resolves to 0 or more than 1 {@link Integration} then `undefined` is returned.
+ *
+ * **Note**: This function is an intermediate helper until we migrate the interpreters to be more general
+ * (likely by migrating to top-down algorithms, see https://github.com/functionless/functionless/issues/374#issuecomment-1203313604)
+ *
+ * @param node the node to resolve the {@link Integration} of.
+ * @returns the ONLY {@link Integration} that {@link node} can resolve to, otherwise `undefined`.
+ */
+export function tryFindIntegration(
+  node: FunctionlessNode
+): Integration | undefined {
+  const integrations = tryFindIntegrations(node);
+  if (integrations?.length === 1) {
+    return integrations[0];
+  }
+  return undefined;
+}
+
+/**
+ * A bottom-up algorithm that determines all of the possible {@link Integration}s that a {@link node}
+ * may resolve to at runtime.
+ *
+ * ```ts
+ * declare const table1;
+ * declare const table2;
+ *
+ * const tables = [table1, table2];
+ *
+ * const a = table1;
+ *    // ^ [table1]
+ *
+ * for (const a of tables) {
+ *   const b = a;
+ *          // ^ [table1, table2]
+ *
+ *   const { appsync: { getItem } } = a;
+ *                     // ^ [ table1.appsync.getItem, table2.appsync.getItem ]
+ * }
+ *
+ * const a = tables[0];
+ *          // ^ [table1]
+ *
+ * const { appsync: { getItem } } = table[0];
+ *                   // ^ [ table1.appsync.getItem ]
+ *
+ * const { appsync: { getItem = table1.appsync.getItem } } = table[2];
+ *                   // ^ [ table1.appsync.getItem ] (because of initializer)
+ * ```
+ *
+ * @param node the node to resolve the possible {@link Integration}s of.
+ * @returns a list of all the {@link Integration}s that the {@link node} could evaluate to.
+ */
+export function tryFindIntegrations(node: FunctionlessNode): Integration[] {
+  return resolve(node, undefined).filter(isIntegration);
+
+  /**
+   * Resolve all of the possible values that {@link node} may resolve to at runtime.
+   *
+   * @param node
+   * @param defaultValue default value to use if the value cannot be resolved (set by default initializers in BindingElement)
+   * @returns an array of all the values the {@link node} resolves to.
+   */
+  function resolve(
+    node: FunctionlessNode | undefined,
+    defaultValue: FunctionlessNode | undefined
+  ): any[] {
+    if (node === undefined) {
+      if (defaultValue === undefined) {
+        return [];
+      } else {
+        return resolve(defaultValue, undefined);
+      }
```

Comment:
This seems wrong? Default value can only be determined at runtime.

`const { a = func } = b ? { a: func2 } : { a: undefined }`

`a` is `func` if `b` is `falsey` (runtime) or `func2` is undefined (maybe runtime)

Reply:
We care calling `resolve` on the `defaultValue` which can return 0, 1 or many possible values. In your case, we would have logic for the ConditionExpr that returns `[{a: func2}, {a: undefined}]` which then distributes to `[func2, undefined]` which then unions with `a = func` to create `[func, func2, undefined]` as the set of all possible values that `a` can be.

Reply:
Let's expand on this in https://github.com/functionless/functionless/issues/310 ?